### ISSUE TITLE
Authenticated JSON-RPC endpoint

### ISSUE THREAD

**Issue by AndreMiras** [paradigmxyz/reth#4919]
### Describe the feature

TL;DR:
It would be nice to have an endpoint that exposes exactly what's behind JSON-RPC (8545) but have it under authentication instead.

Full story:
I would like to expose nodes with JSON-RPC (8545) endpoint available but under authentication.
I initially thought leveraging the built-in auth RPC (8551) would be the way to do it as it offers JWT token auth out of the box. Obviously I can setup a reverse proxy that would handle authentication, but ideally I would prefer to avoid that.
However as pointed out by @rkrasiuk and @mattsse on Telegram isn't a good fit because not only auth RPC exposes only a subset of of `eth_` methods (so for instance it doesn't offer `eth_getBalance`), but also it exposes `engine_` methods which we don't want to mess up with.
See spec: https://github.com/ethereum/execution-apis/blob/v1.0.0-beta.3/src/engine/common.md#underlying-protocol
Later we discussed about the possibility to add custom auth support that would guard the existing RPC endpoint.
This would definitely address my requirement so it would be a nice addition.

### Additional context

The Telegram discussion started here https://t.me/paradigm_reth/15965

**Comment by mattsse** [comment]
what kind of auth to you want to use for these?
also jwt?

**Comment by AndreMiras** [comment]
Yes JWT is nice, thanks for asking

**Comment by mattsse** [comment]
marked this as good first issue

we need a few things:

* add cli argument for authentication of http/ws
* add optional JwtSecret field to https://github.com/paradigmxyz/reth/blob/4dceabf06b9ef26422264ad2026f6e23fda38fbb/crates/rpc/rpc-builder/src/lib.rs#L1096-L1096
* add new variant for auth layer:
https://github.com/paradigmxyz/reth/blob/4dceabf06b9ef26422264ad2026f6e23fda38fbb/crates/rpc/rpc-builder/src/lib.rs#L1633-L1639
see authrpc impl for reference:
https://github.com/paradigmxyz/reth/blob/4dceabf06b9ef26422264ad2026f6e23fda38fbb/crates/rpc/rpc-builder/src/auth.rs#L148-L151

**Comment by mattsse** [comment]
@1010adigupta just checking in, are you still working on this? do you need any pointers?

**Comment by 1010adigupta** [comment]
Yes working on it

**Comment by supernovahs** [comment]
All good, Just wanted to know if the assignee is still working on this rn? Thanks

**Comment by 1010adigupta** [comment]
yes

**Comment by 1010adigupta** [comment]
@mattsse  I have done what was asked upto my knowledge, only addition of cli-args for authentication of http/ws is remaining, can you give some pointers on how to do it.
This is my first Reth contribution so, some help will be required.
Thank You
